### PR TITLE
Encode redirect URL if needed on oauth1 callback request

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
The latest BItBucket Server decodes the callback url so that causes `IllegalArgumentException` error. Catch the error and decode the redirect url.

### What issues does this PR fix or reference?

https://issues.redhat.com/browse/CRW-5663

### How to test this PR?

1. Deploy che with the PR image: `quay.io/eclipse/che-server:pr-663`
2. Configure [oauth1 for BitBucket Server](https://eclipse.dev/che/docs/stable/administration-guide/configuring-oauth-1-for-a-bitbucket-server/) instance.
3. Create a workspace from the BitBucket Server repository, accept the oauth agreement.
4. Stop the workspace and revoke the agreement from the oauth application.
5. Start the workspace and accept the agreement again.

See: workspace starts successfully.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth1/OAuthAuthenticationService.java [hunk 1]
```diff
@@ -73,7 +74,14 @@ public Response callback() throws OAuthAuthenticationException, BadRequestExcept
...
     final String providerName = getParameter(parameters, "oauth_provider");
-    final String redirectAfterLogin = getParameter(parameters, "redirect_after_login");
+    String redirectAfterLogin = getParameter(parameters, "redirect_after_login");
+
+    try {
+      URI.create(redirectAfterLogin);
+    } catch (IllegalArgumentException e) {
+      // the redirectUrl was decoded by the CSM provider, so we need to encode it back.
+      redirectAfterLogin = encodeRedirectUrl(redirectAfterLogin);
+    }
```
Comment (ibuziuk) [inline wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth1/OAuthAuthenticationService.java:84]: can we extract this logic in a separate method with a javadoc reference?

Comment (vinokurig) [reply wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth1/OAuthAuthenticationService.java]: done

**Review by ibuziuk — APPROVED** [review]

**Comment by artaleks9** [comment]
@vinokurig 
The bug which is described in the https://github.com/eclipse/che/issues/22871 is reproduced with Bitbucket server OAuth1. So this blocks checking this PR.

**Comment by vinokurig** [comment]
Merging as it was verified by @artaleks9 in the 7.82.x branch: https://github.com/eclipse-che/che-server/pull/664

**Pull request merged**